src/mlpack/core/data/dictionary_encoding_impl.hpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file dictionary_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICTIONARY_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICTIONARY_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dictionary_encoding.hpp"
+
+namespace mlpack {
+namespace data {
+
+void DictionaryEncoding::Reset()
+{
+  mappings.clear();
+  originalStrings.clear();
+}
+
+DictionaryEncoding::DictionaryEncoding(const DictionaryEncoding& oldObject) :
+    originalStrings(oldObject.originalStrings)
+{
+  std::deque<std::string>::iterator jt = originalStrings.begin();
+  for (auto it = oldObject.originalStrings.begin(); it !=
+      oldObject.originalStrings.end(); it++)
+  {
+    mappings[*jt] = oldObject.mappings.at(*it);
+    jt++;
+  }
+}
+
+DictionaryEncoding DictionaryEncoding::operator= (const
+    DictionaryEncoding &oldObject)
+{
+  if (this != &oldObject)
+  {
+    mappings.clear();
+    originalStrings.clear();
+    originalStrings = oldObject.originalStrings;
+    std::deque<std::string>::iterator jt = originalStrings.begin();
+    for (auto it = oldObject.originalStrings.begin(); it !=
+        oldObject.originalStrings.end(); it++)
+    {
+      mappings[*jt] = oldObject.mappings.at(*it);
+      jt++;
+    }
+  }
+}
+
+template<typename TokenizerType>
+void DictionaryEncoding::CreateMap(std::string& input,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(input);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabel = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType>
+void DictionaryEncoding::Encode(const std::vector<std::string>& input,
+                                MatType& output, TokenizerType tokenizer)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  size_t colSize = 0;
+  for (size_t i = 0; i < input.size(); i++)
+  {
+    strView = input[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      token = tokenizer(strView);
+    }
+    colSize = std::max(colSize, dataset[i].size());
+  }
+  size_t curLabel = mappings.size() + 1;
+  output.zeros(dataset.size(), colSize);
```

Comment:
Does `zeros()` handle zero arguments correctly?

Reply:
Sorry, could you please clarify me a bit on this, I didn't get that.

Reply:
I am not sure that you can pass zero as an argument to `zeros()`. Actually, it doesn't have any sense since the matrix can't contain zero rows or zero columns.

Reply:
@lozhnikov , So do I have to add a check ?

Reply:
Does the documentation say anything? If this case is not mentioned in the documentation, I guess it's better to check that `colSize != 0` and `dataset.size() != 0`.

Reply:
I couldn't find anything in documentation , But I did check the code by passing an input array which is empty , to make `dataset.size()==0` and `colSize==0`, but the code didn't break, checked using valgrind too. So i guess, It is not important to add a check here, Since armadillo has armacheck which i guess takes care of this.

Here is the result :
![Screenshot from 2019-07-03 17-42-35](https://user-images.githubusercontent.com/21286134/60590960-f1ef9900-9dba-11e9-8057-55075f7de2c6.png)

Reply:
Yes, I agree.